game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java

```diff
@@ -0,0 +1,526 @@
+
+    final FiringGroup firingGroup = new FiringGroup("", "", List.of(), targetUnits);
+
+    final FireRoundState fireRoundState = new FireRoundState();
+    fireRoundState.setDice(mock(DiceRoll.class));
+
+    final SelectCasualties selectCasualties =
+        new SelectCasualties(
+            battleState,
+            BattleState.Side.OFFENSE,
+            firingGroup,
+            fireRoundState,
+            (arg1, arg2) -> new CasualtyDetails());
+
+    final SelectMainBattleCasualties.Select selectFunction =
+        mock(SelectMainBattleCasualties.Select.class);
+    final CasualtyDetails expected = new CasualtyDetails(targetUnits, List.of(), false);
+    when(selectFunction.apply(any(), any(), anyCollection(), anyInt())).thenReturn(expected);
+
+    final CasualtyDetails details =
+        new SelectMainBattleCasualties(selectFunction).apply(delegateBridge, selectCasualties);
+
+    assertThat(details.getKilled().toArray(), is(targetUnits.toArray()));
+    // edit mode always sets auto calculated to true
+    assertThat(details.getAutoCalculated(), is(true));
+
+    verify(selectFunction)
+        .apply(
+            eq(delegateBridge),
+            eq(selectCasualties),
+            (Collection<Unit>) argThat(containsInAnyOrder(targetUnits.toArray())),
+            eq(0));
+  }
+
+  @Nested
+  class TransportCasualtiesNotRestricted {
+
+    @Test
+    void moreHitPointsThanHits() {
+      final List<Unit> targetUnits = List.of(givenAnyUnit(), givenAnyUnit());
+
+      targetUnits.forEach(
+          unit -> {
+            final UnitAttachment unitAttachment =
+                (UnitAttachment) unit.getType().getAttachment(UNIT_ATTACHMENT_NAME);
+            when(unitAttachment.getHitPoints()).thenReturn(2);
+          });
+
+      final FiringGroup firingGroup = new FiringGroup("", "", List.of(), targetUnits);
+
+      final FireRoundState fireRoundState = new FireRoundState();
+      fireRoundState.setDice(givenDiceRollWithHits(3));
+
+      final SelectCasualties selectCasualties =
+          new SelectCasualties(
+              battleState,
+              BattleState.Side.OFFENSE,
+              firingGroup,
+              fireRoundState,
+              (arg1, arg2) -> new CasualtyDetails());
+
+      final SelectMainBattleCasualties.Select selectFunction =
+          mock(SelectMainBattleCasualties.Select.class);
+      final CasualtyDetails expected = new CasualtyDetails();
+      when(selectFunction.apply(any(), any(), anyCollection(), anyInt())).thenReturn(expected);
+
+      final CasualtyDetails details =
+          new SelectMainBattleCasualties(selectFunction).apply(delegateBridge, selectCasualties);
+
+      assertThat(details, is(expected));
+
+      verify(selectFunction)
+          .apply(
+              eq(delegateBridge),
+              eq(selectCasualties),
+              (Collection<Unit>) argThat(containsInAnyOrder(targetUnits.toArray())),
+              eq(3));
```

Comment:
side-note, it does look like we are doing a pretty heavy lift in this test by verifying further implementation calls. If we can have the code under test instead return a result object, then have the caller invoke the select function, it would be more testable, functional, and likely easier to understand when reading the prod code (fewer side effects)